slatedb/src/retention_iterator.rs

```diff
@@ -72,20 +77,30 @@ impl<T: KeyValueIterator> RetentionIterator<T> {
         retention_timeout: Option<Duration>,
         retention_min_seq: Option<u64>,
         filter_tombstone: bool,
+        sequence_tracker: Arc<SequenceTracker>,
     ) -> BTreeMap<Reverse<u64>, RowEntry> {
         let mut filtered_versions = BTreeMap::new();
+        let now = system_clock.now().timestamp_millis();
         for (idx, (_, entry)) in versions.into_iter().enumerate() {
             // always keep the latest version (idx == 0), for older versions, check if they are within retention window.
             // retention window is defined by either timeout or seq, or both. if both are not set, only the latest version
             // is kept.
             let in_retention_window_by_time = retention_timeout
                 .map(|timeout| {
-                    let create_ts = entry
-                        .create_ts
-                        .expect("a record with no create_ts should not happen");
-                    // TODO: This is wrong! We're mixing logical (create_ts) and physical (system_clock) timestamps.
+                    let create_sys_ts = sequence_tracker
+                        // we use RoundUp so that we are less aggressive about filtering
+                        // rows (e.g. if the requested window is 10min and the closes recorded
+                        // sequence numbers are 9m30s and 10m30s ago for upper/lower rounding
+                        // respectively we will conservatively select to keep this record in
+                        // the filtered results)
+                        .find_ts(entry.seq, FindOption::RoundUp)
+                        .map(|ts| ts.timestamp_millis())
+                        // if the sequence number is greater than the last recorded sequence
+                        // number we just assume that it was produced now (so it effectively
+                        // should be kept in the filtered results)
+                        .unwrap_or(now);
```

Comment:
Seems risky to use `now` vs. `current_system_ts`. Want to drop `now` completely, and movec `current_system_ts` above this line so we can use it in both places?

Reply:
ah 😆 now is `current_system_ts` (I added it before the iteration loop and didn't notice this line right below... I got a little tunnel visioned). I think it actually makes more sense to use `now` the way I have it so that the entire iterator iterates effectively at one timestamp instead of the clock progressing while iteration is happening. WDYT?